File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */


func maxInt(ints []int) int {
    max := 0
    for _, v := range ints {
        if v > max {
            max = v
        }
    }
    return max
}

func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    pathLengths := []int{
        maxDepth(root.Left) + 1,
        maxDepth(root.Right) + 1,
    }
    
    return maxInt(pathLengths)
}

/*
type DepthNode struct {
    Node *TreeNode
    Depth int
}

func maxDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    pathLengths := []int{}
    stack := []*DepthNode{
        {
            Node: root,
            Depth: 1,
        },
    }
    
    for len(stack) > 0 {
        this := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        
        if this.Node.Left == nil && this.Node.Right == nil {
            pathLengths = append(pathLengths, this.Depth)
            continue
        }
        
        if this.Node.Left != nil {
            left := &DepthNode{
                Node: this.Node.Left, 
                Depth: this.Depth+1,
            }
            stack = append(stack,  left)
        }
        
        if this.Node.Right != nil {
            right := &DepthNode{
                Node: this.Node.Right, 
                Depth: this.Depth+1,
            }
            stack = append(stack,  right)
        }
    }
    
    return maxInt(pathLengths)
}
*/

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if root == None:
            return 0
        
        path_lengths = []
        stack = [(root, 1)]
        while stack:
            node, depth = stack.pop()
            
            if node.left == None and node.right == None:
                path_lengths.append(depth)
                continue
                
            if node.left:
                stack.append((node.left, depth+1))
            
            if node.right:
                stack.append((node.right, depth+1))
        
        return max(path_lengths)
"""   
    
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        return self.recursiveHelper(root, 0)

    def recursiveHelper(self, node: Optional[TreeNode], depth) -> int:
        if node == None:
            return depth
        
        return max(self.recursiveHelper(node.left, depth+1), self.recursiveHelper(node.right, depth+1))